// vehicle-config/core/car-model.js
/*<copyright>
Copyright (c) 2012, Motorola Mobility LLC.
All Rights Reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice,
  this list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.

* Neither the name of Motorola Mobility LLC nor the names of its
  contributors may be used to endorse or promote products derived from this
  software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.
</copyright>*/
var Montage = require("montage").Montage;

exports.CarModel = Montage.create(Montage, {

    moduleName: {
        value: null
    },

    manufacturer: {
        value: null
    },

    model: {
        value: null
    },

    year: {
        value: null
    },

    name: {
        dependencies: ["year", "model"],
        get: function() {
            return this.year + " " + this.model;
        }
    },

    description: {
        value: null
    },

    options: {
        value: null
    },

    licensePlateValue: {
        value: null
    },

    basePrice: {
        value: 0
    },

    configuredPrice: {
        dependencies: ["basePrice"],
        get: function() {
            var keys = Object.keys(this.options),
                self = this;

            return keys.reduce(function(subtotal, optionKey) {

                var option = self[optionKey];

                if (!option) {
                    return subtotal;
                } else if (Array.isArray(option)) {
                    //TODO I don't like how the "chosen" concept leaks from BooleanConfigurationOption
                    return subtotal + option.reduce(function(sub, opt) { return opt.chosen ? sub + opt.cost : sub}, 0);
                } else {
                    return subtotal + option.cost;
                }
            }, this.basePrice);
        }
    },

    didCreate: {
        value: function() {
            var keys = Object.keys(this.options),
                keyCount = keys.length;

            while (keyCount) {
                keyCount--;
                // TODO this doesn't correctly pick up on changes to a cost property if this[keys[keyCount]]
                // is an array i.e. the members of that array aren't observed for changes at "cost"
                Montage.addDependencyToProperty(this, keys[keyCount] + ".cost", "configuredPrice");
                console.log(keys[keyCount] + ".cost")
            }

        }
    },

    performConfigurationChange: {
        value: function(configurationProperty) {

            var modelChanged = this.configureModel(configurationProperty),
                configEvent;

            if (modelChanged) {

                this.toHash();

                configEvent = document.createEvent("CustomEvent");
                configEvent.initCustomEvent("configuration", true, true, {configurationProperty: configurationProperty});
                this.dispatchEvent(configEvent, true, true);
            }
        }
    },

    _modelRepresentation: {
        value: null
    },

    modelRepresentation: {
        get: function() {
            return this._modelRepresentation;
        },
        set: function(value) {
            if (value === this._modelRepresentation) {
                return;
            }

            if (this._modelRepresentation) {
                this._modelRepresentation.removeEventListener("materialReady", this, false);
            }

            this._modelRepresentation = value;

            if (this._modelRepresentation) {
                if (this._modelRepresentation.hasMaterials) {
                    this.initializeModelRepresentation();
                } else {
                    this._modelRepresentation.addEventListener("materialReady", this, false);
                }
            }
        }
    },

    handleMaterialReady: {
        value: function() {
            this.initializeModelRepresentation();
        }
    },

    initializeModelRepresentation: {
        value: function() {

        }
    },

    configureModel: {
        value: function(changedProperty) {

        }
    },

    toHash: {
        value: function() {

        }
    },

    fromHash: {
        value: function(hash) {

        }
    }

});
