// m-car/configurator.reel/configurator.js
/* <copyright>
Copyright (c) 2012, Motorola Mobility LLC.
All Rights Reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

* Redistributions of source code must retain the above copyright notice,
  this list of conditions and the following disclaimer.

* Redistributions in binary form must reproduce the above copyright notice,
  this list of conditions and the following disclaimer in the documentation
  and/or other materials provided with the distribution.

* Neither the name of Motorola Mobility LLC nor the names of its
  contributors may be used to endorse or promote products derived from this
  software without specific prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
POSSIBILITY OF SUCH DAMAGE.
</copyright> */
var Montage = require("montage").Montage,
    Component = require("montage/ui/component").Component,
    ArrayController = require("montage/ui/controller/array-controller").ArrayController,
    Popup = require("montage/ui/popup/popup.reel").Popup;

exports.Configurator = Montage.create(Component, {

    bodyColorSection: {
        value: null
    },

    tridionColorSection: {
        value: null
    },

    interiorSection: {
        value: null
    },

    _vehicle: {
        value: null
    },

    vehicle: {
        get: function() {
            return this._vehicle;
        },
        set: function(value) {
            if (value === this._vehicle) {
                return;
            }

            this._vehicle = value;
        }
    },

    bodyColorController: {
        value: null
    },

    tridionColorController: {
        value: null
    },

    wheelsetController: {
        value: null
    },

    interiorController: {
        value: null
    },

    interiorOptionsController: {
        value: null
    },

    didCreate: {
        value: function() {

            this.bodyColorController = ArrayController.create();
            Object.defineBinding(this.bodyColorController, "content", {
                boundObject: this,
                boundObjectPropertyPath: "vehicle.options.bodyColor",
                oneway: true
            });
            Object.defineBinding(this.bodyColorController, "selectedObjects.0", {
                boundObject: this,
                boundObjectPropertyPath: "vehicle.bodyColor"
            });

            this.tridionColorController = ArrayController.create();
            Object.defineBinding(this.tridionColorController, "content", {
                boundObject: this,
                boundObjectPropertyPath: "vehicle.options.tridionColor",
                oneway: true
            });
            Object.defineBinding(this.tridionColorController, "selectedObjects.0", {
                boundObject: this,
                boundObjectPropertyPath: "vehicle.tridionColor"
            });

            this.wheelsetController = ArrayController.create();
            Object.defineBinding(this.wheelsetController, "content", {
                boundObject: this,
                boundObjectPropertyPath: "vehicle.options.wheelset",
                oneway: true
            });
            Object.defineBinding(this.wheelsetController, "selectedObjects.0", {
                boundObject: this,
                boundObjectPropertyPath: "vehicle.wheelset"
            });

            this.interiorController = ArrayController.create();
            Object.defineBinding(this.interiorController, "content", {
                boundObject: this,
                boundObjectPropertyPath: "vehicle.options.interior",
                oneway: true
            });
            Object.defineBinding(this.interiorController, "selectedObjects.0", {
                boundObject: this,
                boundObjectPropertyPath: "vehicle.interior"
            });

            this.interiorOptionsController = ArrayController.create();
            Object.defineBinding(this.interiorOptionsController, "content", {
                boundObject: this,
                boundObjectPropertyPath: "vehicle.options.interiorOptions",
                oneway: true
            });

            window.foo = this;
        }
    },

    prepareForDraw: {
        value: function() {
            //TODO this is a hack to get the selected to draw properly on the first pass
            this.bodyColorController.selectedObjects = [this.vehicle.bodyColor];
            this.tridionColorController.selectedObjects = [this.vehicle.tridionColor];
            this.interiorController.selectedObjects = [this.vehicle.interior];
            this.wheelsetController.selectedObjects = [this.vehicle.wheelset];

            this.customColorSelected = (this.vehicle.customBodyColor === this.vehicle.bodyColor);
            this.vehicle.addPropertyChangeListener("bodyColor", this, false);
        }
    },

    customColorSelected: {
        value: false
    },

    handleChange: {
        value: function(notification) {
            if ("bodyColor" === notification.propertyPath) {
                this.customColorSelected = (this.vehicle.customBodyColor === this.vehicle.bodyColor);
            }
        }
    },

    pickFromModel: {
        value: function(details) {
            //TODO rely on vehicle name/property mapping
            var materialMap = {
                material_14: "bodyColor",
                material_5: "tridionColor",
                interior1: "interior",
                material_19: "wheelset"
                }

            var property = materialMap[details.material.name];

            console.log(property, this.vehicle[property])

            if (property) {
                this.vehicle.highlightModel(property, details.material.name);
                this.highlightSection(property);
            }
        }
    },

    _highlightTimeouts: {
        distinct: true,
        value: {}
    },

    _highlightedSections: {
        distinct: true,
        value: {}
    },

    highlightDuration: {
        value: 400
    },

    highlightSection: {
        value: function(property) {
            clearTimeout(this._highlightTimeouts[property]);

            this._highlightedSections[property] = true;
            this.needsDraw = true;

            var self = this;
            this._highlightTimeouts = setTimeout(function() {
                self.unhighlightSection(property);
            }, this.highlightDuration);
        }
    },

    unhighlightSection: {
        value: function(property) {
            this._highlightedSections[property] = false;
            this.needsDraw = true;
        }
    },

    draw: {
        value: function() {

            var keys = Object.keys(this._highlightedSections)
            var keyCount = keys.length;

            while(keyCount) {
                keyCount--;
                var property = keys[keyCount];
                var sectionElement = this[property + "Section"];

                if (!sectionElement) {
                    continue;
                }

                var highlighted = this._highlightedSections[property];

                if (highlighted) {
                    sectionElement.classList.add("highlighted");
                } else {
                    sectionElement.classList.remove("highlighted");
                }
            }


        }
    }


});
